### DISCUSSION TITLE
CMAKE_BUILD_TYPE Should be Set to build_type When Using Variants

### DISCUSSION THREAD

**Discussion by daeden** [microsoft/vscode-cmake-tools#2249] [General]
Hello again!

I've started using variants somewhat recently and it is a great feature! My only issue is that the CMAKE_BUILD_TYPE is not being set to build_type and instead is being set to the full variant.

For example, this is what I have in my cmake-variants.yaml

build_type:
  default: debug
  choices:
    debug:
      short: Debug
      build_type: Debug
      settings:
        _build_type: Debug
    release:
      short: Release
      build_type: Release
      settings:
        _build_type: Release

cts_product:
  default: all
  choices:
    all:
      short: all
      settings:
        _cts_product: all

build_list:
  default: cts
  choices:
    cts:
      short: cts
      settings:
        _build_list:
          - crabel-shared
          - core
          - cts
          - nodes
    qra:
      short: qra
      settings:
        _build_list:
          - crabel-shared
          - qra
 

As you can see I set built_type as either 'Debug' or 'Release'. But when I build I see the following:

[build] Build finished with exit code 0

[cpptools] The build configurations generated do not contain the active build configuration. Using 'Debug' for CMAKE_BUILD_TYPE instead of 'Release all qra' to ensure that IntelliSense configurations can be found

[cpptools] The build configurations generated do not contain the active build configuration. Using 'Debug' for CMAKE_BUILD_TYPE instead of 'Release all qra' to ensure that IntelliSense configurations can be found

At first I thought this was not that big a deal, but recall that I'm using buildType to control my build directory!

https://github.com/microsoft/vscode-cmake-tools/discussions/2208

Anyway, this seems like a bug maybe? Or am I doing something wrong?

**Comment by bobbrow** [comment]
This is tracked in #2120 and there is a PR out for it that we need to test before releasing a version with the fix.  We're not sure how long this has been broken but it manifested itself in confusing ways which is why we made the change to log a warning and override the IntelliSense configuration temporarily.

The warning is only about IntelliSense configuration and can be safely ignored until we release a fix.

**Reply by daeden** [reply]
I thought it was benign, but this seems to impact the ${buildType} variable. Unless something else is going on, the issue is not just a warning.

Before I switched to using variants, switching between Debug and Release meant that my build directory would change from buildDebug to buildRelease.

I accomplish this by having:

    "cmake.buildDirectory": "${workspaceFolder}/build${buildType}"

But now the build directory is stubbornly always buildDebug. Perhaps the cpptools warning is actually unrelated. If so, do you how I can fix this issue? It is a big time sink to have to recompile everything when switching from debug to release :(

**Reply by bobbrow** [reply]
You're right.  I was starting a follow up reply that it looked like the actual configuration was incorrect too (which we hadn't seen in previous reports), but had to switch over to some meetings before I finished.

**Reply by bobbrow** [reply]
I created an issue for this.  #2252

**Reply by bobbrow** [reply]
Can you try using `buildType` instead of `build_type`? I don't think the latter is linked to `CMAKE_BUILD_TYPE`.

**Reply by daeden** [reply]
OMG! I can't believe my eyes did not see that I was using build_type instead of buildType!

Thanks Bob! I'm back!

And as always, thank you for the quick response!
